File: 2163_josephus2.js
const fs = require("fs");

let input = fs.readFileSync(0, "utf-8").trim();
let [n, k] = input.split(" ").map(Number);

let result = getJosephusForN(n, k);
console.log(result.join(" "));

function getJosephusForN(n, k) {
    let arr = Array.from({ length: n }, (_, i) => i + 1);
    return josephus(arr, k);
}

function josephus(arr, k) {
    if (!arr) return [];
    if (k == 0) return arr;
    if (arr.length == 1) return arr;
    if (arr.length == 2) return k % 2 == 0 ? arr : arr.slice().reverse();
    else {
        let n = arr.length;
        let idx = k % n;
        const removed = arr[idx];
        const before = arr.slice(0, idx);
        const after = arr.slice(idx + 1);
        const remaining = after.concat(before);
        return [removed].concat(josephus(remaining, k));
    }
}

File: 2163_josephus2.hs
import Data.List (intercalate)

getNext :: [Int] -> Int -> [Int]
getNext [] _ = []
getNext xs 0 = xs
getNext [x] _ = [x]
getNext [x,y] k = if even k then [x,y] else [y,x]
getNext xs k = 
        let 
            n = length xs
            idx = k `mod` n
            (before, after) = splitAt idx xs
            removed = head after
            remaining = (tail after) ++ before
        in removed : getNext remaining k

getResult :: Int -> Int -> [Int]
getResult n k = getNext [1..n] k

main :: IO()
main = do
    line <- getLine
    let [n, k] = map read (words line)
    let [n, k,_] = map read (words line)
    let result_arr = getResult n k 
    let result = intercalate " " (map show result_arr)
    putStrLn result




